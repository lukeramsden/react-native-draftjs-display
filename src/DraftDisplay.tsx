import React, {createContext} from 'react';
import {View} from 'react-native';
import type {
  RawDraftContentState,
  DraftBlockType,
  RawDraftContentBlock,
} from 'draft-js';
import DraftText from './DraftText';

export interface Props {
  contentState: RawDraftContentState;
  blockHandlers?: Record<
    DraftBlockType,
    (props: {block: RawDraftContentBlock}) => React.ReactElement
  >;
}

export const DraftDisplayContext = createContext<{
  getContentState?: () => RawDraftContentState;
}>({});

/**
 * Top-level element
 * Takes contentState and renders it according to configuration
 * contentState is provided to sub-components via React context
 */
const DraftDisplay: React.FC<Props> = ({contentState, blockHandlers}) => {
  return (
    <DraftDisplayContext.Provider value={{getContentState: () => contentState}}>
      {contentState.blocks.map((block) => {
        const BlockHandler = blockHandlers?.[block.type];
        if (BlockHandler) {
          return (
            <View key={block.key}>
              <BlockHandler block={block} />
            </View>
          );
        }
        switch (block.type) {
          case 'value':
          case 'unstyled':
          case 'paragraph':
          case 'header-one':
          case 'header-two':
          case 'header-three':
          case 'header-four':
          case 'header-five':
          case 'header-six':
          case 'code-block':
            return (
              <View key={block.key}>
                <DraftText block={block} />
              </View>
            );

          default:
            break;
        }
      })}
    </DraftDisplayContext.Provider>
  );
};

export default DraftDisplay;
