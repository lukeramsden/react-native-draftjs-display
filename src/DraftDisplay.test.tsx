import 'react-native';

import React from 'react';
import renderer from 'react-test-renderer';

import DraftDisplay from './DraftDisplay';
import {RawDraftContentState} from 'draft-js';

it('renders core block and inline styles correctly', () => {
  const contentState: RawDraftContentState = {
    blocks: [
      {
        key: '34b77',
        data: {
          textAlign: 'center',
          textAlignment: 'center',
        },
        text: 'Centered header',
        type: 'header-one',
        depth: 0,
        entityRanges: [],
        inlineStyleRanges: [],
      },
      {
        key: 'b62v5',
        data: {
          textAlign: 'right',
          textAlignment: 'right',
        },
        text: 'Right header',
        type: 'header-two',
        depth: 0,
        entityRanges: [],
        inlineStyleRanges: [],
      },
      {
        key: '9p9a5',
        data: {
          textAlign: 'justify',
        },
        text: 'Justified header',
        type: 'header-three',
        depth: 0,
        entityRanges: [],
        inlineStyleRanges: [],
      },
      {
        key: '9kknd',
        data: {
          textAlign: 'left',
        },
        text: 'Left header',
        type: 'header-three',
        depth: 0,
        entityRanges: [],
        inlineStyleRanges: [],
      },
      {
        key: '516td',
        data: {},
        text: '',
        type: 'header-three',
        depth: 0,
        entityRanges: [],
        inlineStyleRanges: [],
      },
      {
        key: '3h18a',
        data: {},
        text: 'Bold italic strikethrough underline all',
        type: 'unstyled',
        depth: 0,
        entityRanges: [],
        inlineStyleRanges: [
          {
            style: 'BOLD',
            length: 4,
            offset: 0,
          },
          {
            style: 'BOLD',
            length: 3,
            offset: 36,
          },
          {
            style: 'ITALIC',
            length: 6,
            offset: 5,
          },
          {
            style: 'ITALIC',
            length: 3,
            offset: 36,
          },
          {
            style: 'STRIKETHROUGH',
            length: 13,
            offset: 12,
          },
          {
            style: 'STRIKETHROUGH',
            length: 3,
            offset: 36,
          },
          {
            style: 'UNDERLINE',
            length: 9,
            offset: 26,
          },
          {
            style: 'UNDERLINE',
            length: 3,
            offset: 36,
          },
        ],
      },
    ],
    entityMap: {},
  };

  const tree = renderer
    .create(<DraftDisplay contentState={contentState} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
