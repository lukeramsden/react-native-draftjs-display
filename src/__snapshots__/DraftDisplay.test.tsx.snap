// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders core block and inline styles correctly 1`] = `
Array [
  <View>
    <Text
      style={
        Array [
          Object {
            "fontSize": 32,
            "fontWeight": "bold",
            "marginBottom": 8,
            "marginTop": 10,
          },
          Object {
            "textAlign": "center",
          },
        ]
      }
    >
      Centered header
    </Text>
  </View>,
  <View>
    <Text
      style={
        Array [
          Object {
            "fontSize": 24,
            "fontWeight": "bold",
            "marginBottom": 7,
            "marginTop": 9,
          },
          Object {
            "textAlign": "right",
          },
        ]
      }
    >
      Right header
    </Text>
  </View>,
  <View>
    <Text
      style={
        Array [
          Object {
            "fontSize": 19,
            "fontWeight": "bold",
            "marginBottom": 6,
            "marginTop": 8,
          },
          Object {
            "textAlign": "justify",
          },
        ]
      }
    >
      Justified header
    </Text>
  </View>,
  <View>
    <Text
      style={
        Array [
          Object {
            "fontSize": 19,
            "fontWeight": "bold",
            "marginBottom": 6,
            "marginTop": 8,
          },
          Object {
            "textAlign": "left",
          },
        ]
      }
    >
      Left header
    </Text>
  </View>,
  <View>
    <Text
      style={
        Array [
          Object {
            "fontSize": 19,
            "fontWeight": "bold",
            "marginBottom": 6,
            "marginTop": 8,
          },
          undefined,
        ]
      }
    >
      
    </Text>
  </View>,
  <View>
    <Text
      style={
        Array [
          Object {
            "fontSize": 14,
            "fontWeight": "normal",
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Array [
              Object {
                "fontWeight": "bold",
              },
            ],
          ]
        }
      >
        Bold
      </Text>
      <Text
        style={
          Array [
            Array [],
          ]
        }
      >
         
      </Text>
      <Text
        style={
          Array [
            Array [
              Object {
                "fontStyle": "italic",
              },
            ],
          ]
        }
      >
        italic
      </Text>
      <Text
        style={
          Array [
            Array [],
          ]
        }
      >
         
      </Text>
      <Text
        style={
          Array [
            Array [
              Object {
                "textDecorationLine": "line-through",
              },
            ],
          ]
        }
      >
        strikethrough
      </Text>
      <Text
        style={
          Array [
            Array [],
          ]
        }
      >
         
      </Text>
      <Text
        style={
          Array [
            Array [
              Object {
                "textDecorationLine": "underline",
              },
            ],
          ]
        }
      >
        underline
      </Text>
      <Text
        style={
          Array [
            Array [],
          ]
        }
      >
         
      </Text>
      <Text
        style={
          Array [
            Array [
              Object {
                "fontWeight": "bold",
              },
              Object {
                "fontStyle": "italic",
              },
              Object {
                "textDecorationLine": "line-through",
              },
              Object {
                "textDecorationLine": "underline",
              },
            ],
          ]
        }
      >
        all
      </Text>
    </Text>
  </View>,
]
`;
